impl/src/main/java/org/jboss/weld/util/Proxies.java

```diff
@@ -56,32 +58,37 @@
 
         private final List<Class<?>> interfaces;
         private final List<Class<?>> classes;
+        private final Map<Class<?>, String> classToPackageMap;
 
         private TypeInfo(Set<? extends Type> types) {
             Comparator<Class<?>> classComparator = Comparator.comparing(Class::getName);
             List<Class<?>> foundInterfaces = new ArrayList<>();
             List<Class<?>> foundClasses = new ArrayList<>();
+            Map<Class<?>, String> classToPackage = new HashMap<>();
 
-            types.stream().forEach(type -> add(type, foundInterfaces, foundClasses));
+            types.stream().forEach(type -> add(type, foundInterfaces, foundClasses, classToPackage));
 
             // sort both collections and create unmodifiable lists
             Collections.sort(foundClasses, classComparator);
             Collections.sort(foundInterfaces, classComparator);
+            // TODO should we re-sort this in a way that non-java[x] class names are first?
             this.interfaces = Collections.unmodifiableList(foundInterfaces);
             this.classes = Collections.unmodifiableList(foundClasses);
+            this.classToPackageMap = Collections.unmodifiableMap(classToPackage);
```

Comment:
I think that using Weld immutables would be more effective here, e.g. the list of interfaces can often be empty.